### PR TITLE
metrics: Add reload interval bias metrics for monitor domain reload schema

### PR DESCRIPTION
**Pull request by blacktear23**
### What problem does this PR solve?

Problem Summary:

In heavy load server, we found golang's scheduler will not ensure Timer or select triggered as we expect, so the Timer based logic may get more interval time.

### What is changed and how it works?

What's Changed:

Add domain.Reload called interval metrics can let we know what happened in real heavy load server.

### Related changes

- Need update grafana dashboard configuration.

### Check List 

- Manual test: use `curl` check metrics exists in response.

Side effects

- No

### Release note 

- metrics: Add domain reload interval related metrics.

### CODE REVIEW

**Comment by blacktear23** [comment]
@wjhuang2016 @AilinKid PTAL

File: domain/domain.go [hunk 1]
```diff
@@ -520,6 +546,7 @@ func (do *Domain) loadSchemaInLoop(ctx context.Context, lease time.Duration) {
 	defer util.Recover(metrics.LabelDomain, "loadSchemaInLoop", nil, true)
 	// Lease renewal can run at any frequency.
 	// Use lease/2 here as recommend by paper.
+	logutil.BgLogger().Info(fmt.Sprintf("[DEBUG] Lease time: %v", lease))
```
Comment (xhebox) [inline domain/domain.go:549]: Should it be `.Debug()`?

File: domain/domain.go [hunk 2]
```diff
@@ -358,10 +363,31 @@ func (do *Domain) Reload() error {
...
+			metrics.LoadSchemaIntervalBias.Observe(float64(intervalOffset.Milliseconds()))
+		}
+		if intervalOffset >= do.ddlLease {
+			// more than 2 * ddlLease
+			metrics.LoadSchemaIntervalWarning.WithLabelValues("one_ddl_lease").Inc()
+		} else if intervalOffset > do.ddlLease/2 {
+			// a little more than half ddlLease
+			metrics.LoadSchemaIntervalWarning.WithLabelValues("half_ddl_lease").Inc()
+		}
+		if intervalOffset >= 0 {
```
Comment (xhebox) [inline domain/domain.go:385]: This branch is duplicated with the one at L375?

File: domain/domain.go [hunk 3]
```diff
@@ -358,10 +363,31 @@ func (do *Domain) Reload() error {
...
+		reloadInterval := startTime.Sub(do.lastReloadAt)
+		intervalOffset := reloadInterval - do.ddlLease/2
+		if intervalOffset >= 0 {
+			metrics.LoadSchemaIntervalBias.Observe(float64(intervalOffset.Milliseconds()))
+		}
+		if intervalOffset >= do.ddlLease {
+			// more than 2 * ddlLease
+			metrics.LoadSchemaIntervalWarning.WithLabelValues("one_ddl_lease").Inc()
+		} else if intervalOffset > do.ddlLease/2 {
+			// a little more than half ddlLease
```
Comment (xhebox) [inline domain/domain.go:382]: I am a bit confusing, the first branch means `>= 3/2 * ddlLease`, the 2nd branch means `> ddlLease`. But the comment is not the case?

**Review by xhebox — COMMENTED** [review]
Rest LGTM.

File: domain/domain.go (hunk 3 above)
Comment (blacktear23) [reply domain/domain.go:382]: first branch means greater than a lease, second branch means less than a lease and greater than half lease.

**Comment by blacktear23** [comment]
@xhebox Do you think we need the LoadSchemaIntervalWarning metrics?

File: domain/domain.go (hunk 3 above)
Comment (xhebox) [reply domain/domain.go]: OK, I am not that familiar. So, `intervalOffset := reloadInterval - do.ddlLease/2` is needed and intended, right?

Comment (blacktear23) [reply domain/domain.go]: Yes

**Comment by xhebox** [comment]
> @xhebox Do you think we need the LoadSchemaIntervalWarning metrics?

I don't quite know... I am new to TiDB. IMO, the histogram actually can show warnings, unless `ddlLease` is changing from time to time. I mean, you can check how many counts are in the range `[0, half ddlLease)`.

**Comment by blacktear23** [comment]
> > @xhebox Do you think we need the LoadSchemaIntervalWarning metrics?
> 
> I don't quite know... I am new to TiDB. IMO, the histogram actually can show warnings, unless `ddlLease` is changing from time to time. I mean, you can check how many counts are in the range `[0, half ddlLease)`.

I agreed. @wjhuang2016 @AilinKid what do you think about that?

File: domain/domain.go [hunk 4]
```diff
@@ -358,10 +363,29 @@ func (do *Domain) Reload() error {
 
 	// Lock here for only once at the same time.
 	do.m.Lock()
-	defer do.m.Unlock()
+	defer func() {
+		do.lastReloadAt = time.Now()
```
Comment (zz-jason) [inline domain/domain.go:367]: It's better to set it to `startTime`.

**Review by zz-jason — COMMENTED** [review]
the newly added metrics only show the abnormal phenomena about reload schema intervals. it's better to also show the root cause, do you have any idea? If the only root cause is the high pressure on golang goroutine scheduling, I think it's better to monitor this.

**Comment by blacktear23** [comment]
@zz-jason as I know, this will only happen when Golang's scheduler with high pressure.

File: domain/domain.go (hunk 4 above)
Comment (blacktear23) [reply domain/domain.go:367]: I think lastReloadAt is more meaningful. We are tracing the duration between 2 reload operations, not one reload operation take how many seconds.

Comment (blacktear23) [reply domain/domain.go:367]: Maybe `lastReloadedAt` is better?

**Comment by wjhuang2016** [comment]
Need update grafana dashboard configuration

**Comment by wjhuang2016** [comment]
Need update grafana dashboard configuration

**Comment by blacktear23** [comment]
@wjhuang2016 handwriting the json file? 😂

**Comment by wjhuang2016** [comment]
> @wjhuang2016 handwriting the json file? 😂

Start a grafana, add a panel, then export the JSON.

**Comment by zz-jason** [comment]
I'm going to close this PR since it hasn't been updated for a long time. Feel free to reopen it once you wish to advance this PR in the future.

**Pull request closed without merging**